sellable/vehicle.go

```diff
@@ -0,0 +1,158 @@
+	lib.CheckError(err)
+	product, err := Vehicle(authToken.Role, policy)
+
+	jsonProduct, err := json.Marshal(product)
+	if err != nil {
+		return "", models.Product{}, fmt.Errorf("cannot generate the JSON: %v", err)
+	}
+
+	return string(jsonProduct), product, err
+}
+
+// Given a policy that should contain the Vehicle and the Person assets, then it returns:
+//   - the product or parts of it depending on the sellable rules
+//   - and an eventual error
+func Vehicle(role string, p models.Policy) (models.Product, error) {
+	if err := validatePolicy(p); err != nil {
+		return models.Product{}, fmt.Errorf("the policy did not pass validation: %v", err)
+	}
+
+	if err := isVehicleSellable(p); err != nil {
+		return models.Product{}, fmt.Errorf("vehicle not sellable: %v", err)
+	}
+
+	product, err := productForVehicle(p, role)
+	if err != nil {
+		return models.Product{}, fmt.Errorf("no products for this vehicle: %v", err)
+	}
+
+	return product, nil
+}
+
+func productForVehicle(p models.Policy, r string) (models.Product, error) {
+	product, err := prd.GetProduct("gap", "v1", r)
+	if err != nil {
+		return models.Product{}, fmt.Errorf("error in getting the product: %v", err)
+	}
+
+	vp := p.Assets[0].Vehicle.PriceValue
+	if vp > minPriceOnlyComplete && vp <= maxPriceValue {
+		delete(product.Offers, "base")
+	}
+
+	return product, nil
+}
+
+// Returns true if the policy is conforming to the sellability rules
+func isVehicleSellable(p models.Policy) error {
+	v := p.Assets[0].Vehicle
+	if !v.IsFireTheftCovered {
+		return fmt.Errorf("fire and theft is not covered")
+	}
+	if v.MainUse != "private" {
+		return fmt.Errorf("the vehicle is not private")
+	}
+
+	car_types := []string{"auto", "autocarro", "suv"}
+	if !lib.SliceContains(car_types, v.VehicleType) {
+		return fmt.Errorf("The vehicle type is not in: %v", car_types)
+	}
+
+	anniversary := v.RegistrationDate.AddDate(maxAgeAtStartDate, 0, 0)
+	if p.StartDate.After(anniversary) {
+		return fmt.Errorf("The registration is too old, exceeded the start date")
+	}
+
+	anniversary = v.RegistrationDate.AddDate(maxAgeAtEndDate, 0, 0)
+	if p.EndDate.After(anniversary) {
+		return fmt.Errorf("The registration is too old, exceeded the end date")
+	}
+
+	vp := p.Assets[0].Vehicle.PriceValue
+	if vp < minPriceValue || vp > maxPriceValue {
+		return fmt.Errorf("the value is not within the accepted range")
+	}
+
+	return nil
+}
+
+func validatePolicy(p models.Policy) error {
+	if len(p.Assets) == 0 {
+		return fmt.Errorf("no assets found")
+	}
+
+	if p.Assets[0].Person == nil {
+		return fmt.Errorf("no person found")
+	}
+
+	if p.Assets[0].Vehicle == nil {
+		return fmt.Errorf("no vehicle found")
+	}
+
+	v := p.Assets[0].Vehicle
+	pd := lib.ElapsedYears(p.StartDate, p.EndDate)
+
+	maxAgeFullCoverageBD := v.RegistrationDate.AddDate(maxAgeFullCoverage, 0, 0)
+	if time.Now().Before(maxAgeFullCoverageBD) {
+		if pd != maxCoverage {
+			return fmt.Errorf("wrong policy duration! it should be 5, we've got %d", pd)
+		}
```

Comment:
this should be <= and not != because you can have a duration up to 5 years if the car has less than 3 years old